File: frontend/src/lib/stores/troop.ts
import { writable } from "svelte/store";
import { toast } from "svelte-sonner";
import { api } from "../api/client";

// Troop types matching backend
export type TroopType =
    | 'infantry'
    | 'spearman'
    | 'war_elephant'
    | 'buffalo_wagon'
    | 'kris_warrior'
    | 'sea_diver'
    | 'war_prahu'
    | 'merchant_ship'
    | 'crossbowman'
    | 'mountain_warrior'
    | 'highland_pony'
    | 'trap_maker'
    | 'swamp_dragon'
    | 'locust_swarm'
    | 'battle_duck'
    | 'portuguese_musketeer'
    // Chief units (can reduce loyalty)
    | 'royal_advisor'
    | 'harbor_master'
    | 'elder_chief';

export type TribeType = 'phasuttha' | 'nava' | 'kiri' | 'special';

export interface TroopDefinition {
    troop_type: TroopType;
    tribe: TribeType;
    name: string;
    description: string | null;
    attack: number;
    defense_infantry: number;
    defense_cavalry: number;
    speed: number;
    carry_capacity: number;
    crop_consumption: number;
    training_time_seconds: number;
    wood_cost: number;
    clay_cost: number;
    iron_cost: number;
    crop_cost: number;
    required_building: string;
    required_building_level: number;
    loyalty_reduction: number;
}

export interface Troop {
    troop_type: TroopType;
    count: number;
    in_village: number;
    on_mission: number;
}

export interface TroopQueueItem {
    id: string;
    troop_type: TroopType;
    count: number;
    each_duration_seconds: number;
    started_at: string;
    ends_at: string;
}

export interface TroopCost {
    wood: number;
    clay: number;
    iron: number;
    crop: number;
    time_seconds: number;
}

export interface TrainResponse {
    queue_entry: TroopQueueItem;
    cost: TroopCost;
}

interface TroopState {
    definitions: TroopDefinition[];
    troops: Troop[];
    queue: TroopQueueItem[];
    loading: boolean;
    error: string | null;
}

function createTroopStore() {
    const { subscribe, set, update } = writable<TroopState>({
        definitions: [],
        troops: [],
        queue: [],
        loading: false,
        error: null,
    });

    return {
        subscribe,

        // Load all troop definitions
        loadDefinitions: async () => {
            update(state => ({ ...state, loading: true, error: null }));
            try {
                const definitions = await api.get<TroopDefinition[]>('/api/troops/definitions');
                update(state => ({
                    ...state,
                    definitions,
                    loading: false,
                }));
                return definitions;
            } catch (error: any) {
                const message = error.message || 'Failed to load troop definitions';
                update(state => ({
                    ...state,
                    loading: false,
                    error: message,
                }));
                throw error;
            }
        },

        // Load troops for a village
        loadTroops: async (villageId: string) => {
            update(state => ({ ...state, loading: true, error: null }));
            try {
                const troops = await api.get<Troop[]>(`/api/villages/${villageId}/troops`);
                update(state => ({
                    ...state,
                    troops,
                    loading: false,
                }));
                return troops;
            } catch (error: any) {
                const message = error.message || 'Failed to load troops';
                update(state => ({
                    ...state,
                    loading: false,
                    error: message,
                }));
                throw error;
            }
        },

        // Load training queue for a village
        loadQueue: async (villageId: string) => {
            try {
                const queue = await api.get<TroopQueueItem[]>(`/api/villages/${villageId}/troops/queue`);
                update(state => ({ ...state, queue }));
                return queue;
            } catch (error: any) {
                console.error('Failed to load training queue:', error);
                return [];
            }
        },

        // Train troops
        train: async (villageId: string, troopType: TroopType, count: number) => {
            update(state => ({ ...state, loading: true, error: null }));
            try {
                const response = await api.post<TrainResponse>(
                    `/api/villages/${villageId}/troops/train`,
                    { troop_type: troopType, count }
                );

                update(state => ({
                    ...state,
                    queue: [...state.queue, response.queue_entry],
                    loading: false,
                }));

                toast.success('Training Started', {
                    description: `Training ${count} ${formatTroopName(troopType)}`
                });

                return response;
            } catch (error: any) {
                const message = error.message || 'Failed to start training';
                update(state => ({
                    ...state,
                    loading: false,
                    error: message,
                }));
                toast.error('Training Failed', { description: message });
                throw error;
            }
        },

        // Cancel training
        cancelTraining: async (villageId: string, queueId: string) => {
            update(state => ({ ...state, loading: true, error: null }));
            try {
                await api.delete(`/api/villages/${villageId}/troops/queue/${queueId}`);

                update(state => ({
                    ...state,
                    queue: state.queue.filter(q => q.id !== queueId),
                    loading: false,
                }));

                toast.success('Training Cancelled');

                return true;
            } catch (error: any) {
                const message = error.message || 'Failed to cancel training';
                update(state => ({
                    ...state,
                    loading: false,
                    error: message,
                }));
                toast.error('Cancel Failed', { description: message });
                throw error;
            }
        },

        // Clear error
        clearError: () => {
            update(state => ({ ...state, error: null }));
        },

        // Reset store
        reset: () => {
            set({
                definitions: [],
                troops: [],
                queue: [],
                loading: false,
                error: null,
            });
        },
    };
}

export const troopStore = createTroopStore();

// Helper functions
export function formatTroopName(type: TroopType): string {
    return type
        .split('_')
        .map(word => word.charAt(0).toUpperCase() + word.slice(1))
        .join(' ');
}

export function getTroopIcon(type: TroopType): string {
    const icons: Record<TroopType, string> = {
        infantry: '⚔️',
        spearman: '🗡️',
        war_elephant: '🐘',
        buffalo_wagon: '🛒',
        kris_warrior: '🗡️',
        sea_diver: '🤿',
        war_prahu: '⛵',
        merchant_ship: '🚢',
        crossbowman: '🏹',
        mountain_warrior: '⛰️',
        highland_pony: '🐎',
        trap_maker: '🪤',
        swamp_dragon: '🐊',
        locust_swarm: '🦗',
        battle_duck: '🦆',
        portuguese_musketeer: '🔫',
        // Chief units
        royal_advisor: '👑',
        harbor_master: '⚓',
        elder_chief: '🧙',
    };
    return icons[type] || '👤';
}

export function getTribeColor(tribe: TribeType): string {
    const colors: Record<TribeType, string> = {
        phasuttha: 'text-amber-600',
        nava: 'text-blue-600',
        kiri: 'text-green-600',
        special: 'text-purple-600',
    };
    return colors[tribe] || 'text-gray-600';
}

export function formatTime(seconds: number): string {
    const hours = Math.floor(seconds / 3600);
    const minutes = Math.floor((seconds % 3600) / 60);
    const secs = seconds % 60;

    if (hours > 0) {
        return `${hours}h ${minutes}m`;
    }
    if (minutes > 0) {
        return `${minutes}m ${secs}s`;
    }
    return `${secs}s`;
}

export function getDefinitionByType(definitions: TroopDefinition[], type: TroopType): TroopDefinition | undefined {
    return definitions.find(d => d.troop_type === type);
}

// Chief troops can reduce village loyalty during Conquer missions
export function isChiefTroop(type: TroopType): boolean {
    return type === 'royal_advisor' || type === 'harbor_master' || type === 'elder_chief';
}

File: backend/src/models/troop.rs
use chrono::{DateTime, Utc};
use serde::{Deserialize, Serialize};
use sqlx::FromRow;
use uuid::Uuid;

use super::building::BuildingType;

#[derive(Debug, Clone, Copy, Serialize, Deserialize, sqlx::Type, PartialEq, Eq, Hash)]
#[sqlx(type_name = "troop_type", rename_all = "snake_case")]
#[serde(rename_all = "snake_case")]
pub enum TroopType {
    // Phasuttha (Mainland/Thai-inspired)
    Infantry,
    Spearman,
    WarElephant,
    BuffaloWagon,
    // Nava (Maritime/Malay-inspired)
    KrisWarrior,
    SeaDiver,
    WarPrahu,
    MerchantShip,
    // Kiri (Highland/Hill tribe-inspired)
    Crossbowman,
    MountainWarrior,
    HighlandPony,
    TrapMaker,
    // Special units
    SwampDragon,
    LocustSwarm,
    BattleDuck,
    PortugueseMusketeer,
    // Chief units (can reduce loyalty)
    RoyalAdvisor,
    HarborMaster,
    ElderChief,
}

#[derive(Debug, Clone, Copy, Serialize, Deserialize, sqlx::Type, PartialEq)]
#[sqlx(type_name = "tribe_type", rename_all = "snake_case")]
#[serde(rename_all = "snake_case")]
pub enum TribeType {
    Phasuttha,
    Nava,
    Kiri,
    Special,
}

impl TroopType {
    pub fn tribe(&self) -> TribeType {
        match self {
            TroopType::Infantry | TroopType::Spearman | TroopType::WarElephant | TroopType::BuffaloWagon | TroopType::RoyalAdvisor => TribeType::Phasuttha,
            TroopType::KrisWarrior | TroopType::SeaDiver | TroopType::WarPrahu | TroopType::MerchantShip | TroopType::HarborMaster => TribeType::Nava,
            TroopType::Crossbowman | TroopType::MountainWarrior | TroopType::HighlandPony | TroopType::TrapMaker | TroopType::ElderChief => TribeType::Kiri,
            _ => TribeType::Special,
        }
    }

    pub fn is_cavalry(&self) -> bool {
        matches!(
            self,
            TroopType::WarElephant
                | TroopType::BuffaloWagon
                | TroopType::HighlandPony
                | TroopType::WarPrahu
                | TroopType::MerchantShip
        )
    }

    pub fn is_infantry(&self) -> bool {
        !self.is_cavalry()
    }

    /// Check if this troop type is a Chief (can reduce loyalty)
    pub fn is_chief(&self) -> bool {
        matches!(
            self,
            TroopType::RoyalAdvisor | TroopType::HarborMaster | TroopType::ElderChief
        )
    }
}

/// Troop definition from database (base stats)
#[derive(Debug, Clone, Serialize, Deserialize, FromRow)]
pub struct TroopDefinition {
    pub id: Uuid,
    pub troop_type: TroopType,
    pub tribe: TribeType,
    pub name: String,
    pub description: Option<String>,
    // Combat stats
    pub attack: i32,
    pub defense_infantry: i32,
    pub defense_cavalry: i32,
    pub speed: i32,
    pub carry_capacity: i32,
    pub crop_consumption: i32,
    // Training
    pub training_time_seconds: i32,
    pub wood_cost: i32,
    pub clay_cost: i32,
    pub iron_cost: i32,
    pub crop_cost: i32,
    // Requirements
    pub required_building: BuildingType,
    pub required_building_level: i32,
    // Conquer ability
    pub loyalty_reduction: i32,
    pub created_at: DateTime<Utc>,
}

/// Troops owned by a village
#[derive(Debug, Clone, Serialize, Deserialize, FromRow)]
pub struct Troop {
    pub id: Uuid,
    pub village_id: Uuid,
    pub troop_type: TroopType,
    pub count: i32,
    pub in_village: i32,
    pub created_at: DateTime<Utc>,
    pub updated_at: DateTime<Utc>,
}

/// Training queue entry
#[derive(Debug, Clone, Serialize, Deserialize, FromRow)]
pub struct TroopQueue {
    pub id: Uuid,
    pub village_id: Uuid,
    pub troop_type: TroopType,
    pub count: i32,
    pub each_duration_seconds: i32,
    pub started_at: DateTime<Utc>,
    pub ends_at: DateTime<Utc>,
    pub created_at: DateTime<Utc>,
}

// Request/Response DTOs

#[derive(Debug, Clone, Deserialize)]
pub struct TrainTroopsRequest {
    pub troop_type: TroopType,
    pub count: i32,
}

#[derive(Debug, Clone, Serialize)]
pub struct TrainTroopsResponse {
    pub queue_entry: TroopQueueResponse,
    pub cost: TroopCost,
}

#[derive(Debug, Clone, Serialize)]
pub struct TroopCost {
    pub wood: i32,
    pub clay: i32,
    pub iron: i32,
    pub crop: i32,
    pub time_seconds: i32,
}

#[derive(Debug, Clone, Serialize)]
pub struct TroopResponse {
    pub troop_type: TroopType,
    pub count: i32,
    pub in_village: i32,
    pub on_mission: i32,
}

impl From<Troop> for TroopResponse {
    fn from(t: Troop) -> Self {
        Self {
            troop_type: t.troop_type,
            count: t.count,
            in_village: t.in_village,
            on_mission: t.count - t.in_village,
        }
    }
}

#[derive(Debug, Clone, Serialize)]
pub struct TroopQueueResponse {
    pub id: Uuid,
    pub troop_type: TroopType,
    pub count: i32,
    pub each_duration_seconds: i32,
    pub started_at: DateTime<Utc>,
    pub ends_at: DateTime<Utc>,
}

impl From<TroopQueue> for TroopQueueResponse {
    fn from(q: TroopQueue) -> Self {
        Self {
            id: q.id,
            troop_type: q.troop_type,
            count: q.count,
            each_duration_seconds: q.each_duration_seconds,
            started_at: q.started_at,
            ends_at: q.ends_at,
        }
    }
}

#[derive(Debug, Clone, Serialize)]
pub struct TroopDefinitionResponse {
    pub troop_type: TroopType,
    pub tribe: TribeType,
    pub name: String,
    pub description: Option<String>,
    pub attack: i32,
    pub defense_infantry: i32,
    pub defense_cavalry: i32,
    pub speed: i32,
    pub carry_capacity: i32,
    pub crop_consumption: i32,
    pub training_time_seconds: i32,
    pub wood_cost: i32,
    pub clay_cost: i32,
    pub iron_cost: i32,
    pub crop_cost: i32,
    pub required_building: BuildingType,
    pub required_building_level: i32,
    pub loyalty_reduction: i32,
}

impl From<TroopDefinition> for TroopDefinitionResponse {
    fn from(d: TroopDefinition) -> Self {
        Self {
            troop_type: d.troop_type,
            tribe: d.tribe,
            name: d.name,
            description: d.description,
            attack: d.attack,
            defense_infantry: d.defense_infantry,
            defense_cavalry: d.defense_cavalry,
            speed: d.speed,
            carry_capacity: d.carry_capacity,
            crop_consumption: d.crop_consumption,
            training_time_seconds: d.training_time_seconds,
            wood_cost: d.wood_cost,
            clay_cost: d.clay_cost,
            iron_cost: d.iron_cost,
            crop_cost: d.crop_cost,
            required_building: d.required_building,
            required_building_level: d.required_building_level,
            loyalty_reduction: d.loyalty_reduction,
        }
    }
}
